### PR TITLE
ASoC: intel: sof_sdw: Add CS42L43 CODEC support

### PR DESCRIPTION
**Pull request by bardliao**
Add cs42l43 codec support to sof_sdw machine driver.
The first 6 commits are applied by upstream and the 7th commit is for old codec version only.
Only the last 2 commits need to review.

### CODE REVIEW

File: sound/soc/intel/boards/sof_sdw_cs42l43.c
Comment (charleskeepax) [inline sound/soc/intel/boards/sof_sdw_cs42l43.c]: Should probably drop BTN_4,5

File: sound/soc/intel/boards/sof_sdw_cs42l43.c
Comment (charleskeepax) [inline sound/soc/intel/boards/sof_sdw_cs42l43.c]: Order here should actually be VOICECOMMAND on BTN_1, UP on 2 and DOWN on 3. Apologies that was a bug in our early versions of the machine driver stuff.

File: sound/soc/intel/boards/sof_sdw_cs42l43.c
Comment (charleskeepax) [inline sound/soc/intel/boards/sof_sdw_cs42l43.c]: Slight question on what makes sense here? Does your setup use the Cohen speakers or is it only using Jamerson speakers? Although I guess given the way this stuff works this would be shared across systems so probably should have a superset routing in it.

And if it does use the Cohen ones, at minimum this should be linked to AMP1_OUT_N as well, probably AMP2_OUT_P/N as well assuming both speakers are being used.

File: sound/soc/intel/boards/sof_sdw_cs42l43.c
Comment (charleskeepax) [inline sound/soc/intel/boards/sof_sdw_cs42l43.c]: Should probably add widgets for the headphone, headset mic, on board mics, since we added DAIs for them all.

File: sound/soc/intel/boards/sof_sdw_cs42l43.c
Comment (charleskeepax) [inline sound/soc/intel/boards/sof_sdw_cs42l43.c]: Better to pass a 0 here instead of 4.8MHz, that way Cohen will just use whatever bus speed the SoundWire runs at. Whereas if you pass an exact value it will complain with an error if they differ.

**Review by charleskeepax — CHANGES_REQUESTED** [review]

File: sound/soc/intel/boards/sof_sdw_cs42l43.c
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw_cs42l43.c:50]: this could be fun if we have cohen+jameson connected, we probably need to work on the 'component' string a bit more so that UCM files figure out how the headset codec and amplifiers are setup.

File: sound/soc/intel/boards/sof_sdw_cs42l43.c
Comment (plbossart) [reply sound/soc/intel/boards/sof_sdw_cs42l43.c]: I don't have a good understanding of how this 'sof_sdw' machine driver might handle the Cohen+Jamerson combination, maybe it's an entirely different machine driver really since it's a mix of SPI + SoundWire IIRC?

File: sound/soc/intel/boards/sof_sdw_cs42l43.c
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw_cs42l43.c]: > Should probably drop BTN_4,5

done

File: sound/soc/intel/boards/sof_sdw_cs42l43.c
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw_cs42l43.c]: > Order here should actually be VOICECOMMAND on BTN_1, UP on 2 and DOWN on 3. Apologies that was a bug in our early versions of the machine driver stuff.

done

File: sound/soc/intel/boards/sof_sdw_cs42l43.c
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw_cs42l43.c]: > Better to pass a 0 here instead of 4.8MHz, that way Cohen will just use whatever bus speed the SoundWire runs at. Whereas if you pass an exact value it will complain with an error if they differ.

done

File: sound/soc/intel/boards/sof_sdw_cs42l43.c
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw_cs42l43.c]: On second thought, I think we don't need the map and the widget at all. The purpose of those maps in machine driver is to switch on/off the devices. But we have one pcm per device, the devices will be on/off when we open/close the pcm.

Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw_cs42l43.c]: The configuration is using Cohen's headset and DMIC + Jamerson's speakers. @plbossart In our usage, both Cohen and Jamerson are SoundWire.

File: sound/soc/intel/boards/sof_sdw_cs42l43.c
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw_cs42l43.c:50]: > this could be fun if we have cohen+jameson connected, we probably need to work on the 'component' string a bit more so that UCM files figure out how the headset codec and amplifiers are setup.

I add a separated init callback for DMIC and add `"%s mic:cs42l43-dmic` to card->components.

**Comment by bardliao** [comment]
@plbossart @charleskeepax There is an open in my mind. Currently, I don't add Cohen's speaker DAI in codec_info_list[] because we don't use for now. But, if a customer decides to use Cohen's speaker, we should add it in codec_info_list[].
How should we know if a device uses Cohen's speaker or not? For example,
Device A uses Cohen's Headset, DMIC, and Speaker.
Device B uses Cohen's Headset, DMIC and Jameson's Speaker
Device C uses Cohen's Headset, DMIC, Speaker combines with Jameson's Speaker.
How do we know what exactly a device use?

File: sound/soc/intel/boards/sof_sdw_cs42l43.c
Comment (charleskeepax) [reply sound/soc/intel/boards/sof_sdw_cs42l43.c]: I dont mind skipping them for now, but really machine drivers should generally have representation of the board level audio stuff, like mics and speakers. For example if you mark the card as fully_routed nothing will turn on if its not attached to a proper input/output.

**Comment by charleskeepax** [comment]
> How should we know if a device uses Cohen's speaker or not?

Yeah I don't have a fully formed plan of that either yet, I still have a few things to work through certainly this week maybe slipping into next and then I am probably going to be spending a bit more time on those sorts of issues. For now my best guess is it probably ends up being some sort of quirk.

> I don't have a good understanding of how this 'sof_sdw' machine driver might handle the Cohen+Jamerson combination, maybe it's an entirely different machine driver really since it's a mix of SPI + SoundWire IIRC?

Not all Cohen + Jamerson is that way, most of the systems will be both devices on SoundWire which is simple enough. But some of them will have that odd configuration with the Jamersons hidden behind the Cohen. The current plan there was to add some things to the machine driver to allow creating some codec to codec links on the card, but the patch set isn't really ready for review yet.

File: sound/soc/intel/boards/sof_sdw_cs42l43.c
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw_cs42l43.c]: @charleskeepax Understand. I will add the widgets and the map back. BTW, do you have recommended INPUT/OUTPUT for headset and DMIC? For example, `PDM1_DIN` and `PDM2_DIN` for DMIC.

Comment (charleskeepax) [reply sound/soc/intel/boards/sof_sdw_cs42l43.c]: This is what I currently have internally:

```
static const struct snd_soc_dapm_widget sof_widgets[] = {
        SND_SOC_DAPM_HP("Headphone", NULL),
        SND_SOC_DAPM_MIC("Headset Mic", NULL),
        SND_SOC_DAPM_SPK("Left Speaker", NULL),
        SND_SOC_DAPM_SPK("Right Speaker", NULL),
        SND_SOC_DAPM_MIC("Left Mic", NULL),
        SND_SOC_DAPM_MIC("Right Mic", NULL),
};

static const struct snd_soc_dapm_route sof_map[] = {
        { "Headphone", NULL, "cs42l43 AMP3_OUT" },
        { "Headphone", NULL, "cs42l43 AMP4_OUT" },
        { "cs42l43 ADC1_IN1_P", NULL, "Headset Mic" },
        { "cs42l43 ADC1_IN1_N", NULL, "Headset Mic" },
        { "Left Speaker", NULL, "cs42l43 AMP1_OUT_P" },
        { "Left Speaker", NULL, "cs42l43 AMP1_OUT_N" },
        { "Right Speaker", NULL, "cs42l43 AMP2_OUT_P" },
        { "Right Speaker", NULL, "cs42l43 AMP2_OUT_N" },
        { "cs42l43 PDM2_DIN", NULL, "Left Mic" },
        { "cs42l43 PDM2_DIN", NULL, "Right Mic" },
};
```

Although doing separate left and rights might be excessive, bit of a judgement call there, but certainly happy to drop those if you prefer. I guess that is still missing a route for the stereo microphone jack case, but we can probably ignore that for now as its pretty niche.

On the plus side, I think this routing will be pretty standard everyone seems to be sticking with the connections recommended by the apps guys here.

Comment (charleskeepax) [reply sound/soc/intel/boards/sof_sdw_cs42l43.c]: In fact thinking about it merging left and rights is probably better since it allows you to have a single pin switch control for both if one wishes.

**Comment by bardliao** [comment]
Rebased.
Changes: https://github.com/thesofproject/linux/compare/44a93a6702f2497749ce1d017da570c7f6b56dd3..d04ad53a57ad96499aceb5fda26946a3555717e3

**Review by charleskeepax — DISMISSED** [review]

File: sound/soc/intel/boards/sof_sdw_cs42l43.c
Comment (plbossart) [inline sound/soc/intel/boards/sof_sdw_cs42l43.c:119]: this is a bit weird, it will unconditionally tell userspace/UCM about the presence of dmics attached to the Cirrus Logic codec.

What happens if the dmics are attached to the PCH, as done on Chromebooks? Is this not a supported configuration?

**Review by plbossart — COMMENTED** [review]
question for @charleskeepax

File: sound/soc/intel/boards/sof_sdw_cs42l43.c
Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw_cs42l43.c:119]: > this is a bit weird, it will unconditionally tell userspace/UCM about the presence of dmics attached to the Cirrus Logic codec.
> 
> What happens if the dmics are attached to the PCH, as done on Chromebooks? Is this not a supported configuration?

@plbossart I think the question is how do we know what functions are used in a multi-function codec? Maybe use a mask, but where can we get the mask? Using a dmi quirk? That's something we need to think about. My idea is that we can create dai links based on what functions are used. Then, the card->components will be added only when the function is used.
Btw, I think it will be better if a codec can use different SDW ADR to tell what functions are supported, and customers select the codec that fits their requirement. Like, rt712 supports jack, dmic, and speaker, and rt713 supports jack and dmic. So that we can assume that the speaker will be used when a device uses rt712 instead of rt713.
Maybe using different unique id is a way to tell what functions are used?

Comment (charleskeepax) [reply sound/soc/intel/boards/sof_sdw_cs42l43.c:119]: That would involve OTPing the parts differently would it not? I suspect the hardware folks are very unlikely to go for that plan. The part should have all the disco stuff in ACPI I suspect the best plan might be to parse that to see which bits are being used. I will have a bit of a look and see where I get to on that.

Comment (plbossart) [reply sound/soc/intel/boards/sof_sdw_cs42l43.c:119]: I have limited hopes that _DSD properties would help, most BIOS vendors don't get this right usually. I think all the DisCo plans are vaporware for now, see how well it went for the version 1.0, the only thing we use is the _ADR and even that is wrong on all Intel development devices and some HP/NUC commercial devices derived from the Intel reference.

And anyways the PCH-attached DMIC is not really covered in such properties, it's part of the other Intel NHLT stuff.

When we have a discrete microphone codec present, it's relatively easy to make a decision, but if the Cirrus codec may  or may not have microphones attached then it's more difficult, isn't it?

Comment (charleskeepax) [reply sound/soc/intel/boards/sof_sdw_cs42l43.c:119]: It looks like Windows is making more of a move to use the disco stuff now so that should to some extent help it along, but if we dont want to trust it for now, I think that mostly leaves us with quirks as an option. I guess another option might be to hook up all the interfaces regardless of if they are attached to something although I suppose that doesnt give user-space any info about what it can and cant do.

Comment (plbossart) [reply sound/soc/intel/boards/sof_sdw_cs42l43.c:119]: IIRC we have a restriction in the machine driver, I think we are assuming that either the intel PCH DMIC is used or the SoundWire MIC is used, not both.

Maybe best to check with @bardliao

BTW This isn't a random academic problem I am asking. The RT711 is capable of handling headset and local mics, but we've always used the PCH-DMIC.

Comment (bardliao) [reply sound/soc/intel/boards/sof_sdw_cs42l43.c:119]: > IIRC we have a restriction in the machine driver, I think we are assuming that either the intel PCH DMIC is used or the SoundWire MIC is used, not both.
> 
> Maybe best to check with @bardliao

I think the machine driver can have both SoundWire MIC and PCH DMIC. But, we never have such configuration.

> 
> BTW This isn't a random academic problem I am asking. The RT711 is capable of handling headset and local mics, but we've always used the PCH-DMIC.

Right, but I feel that we will face the issue soon when people want to use Cohen's speaker.

**Review by plbossart — DISMISSED** [review]
we'll have to figure out the DMIC problem at some point, but let's merge this for now.

**Comment by plbossart** [comment]
@bardliao can you rebase and check the build issues?

**Comment by bardliao** [comment]
> @bardliao can you rebase and check the build issues?

Done

**Review by plbossart — APPROVED** [review]

**Comment by plbossart** [comment]
@bardliao I'll let you merge this when @charleskeepax re-adds his review

**Review by charleskeepax — APPROVED** [review]

**Pull request merged**
